src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2018 Demerzel Solutions Limited
+ * This file is part of the Nethermind library.
+ *
+ * The Nethermind library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The Nethermind library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Abi;
+using Nethermind.AuRa.Contracts;
+using Nethermind.Blockchain;
+using Nethermind.Core;
+using Nethermind.Core.Specs.ChainSpecStyle;
+using Nethermind.Core.Specs.Forks;
+using Nethermind.Dirichlet.Numerics;
+using Nethermind.Evm;
+using Nethermind.Evm.Tracing;
+using Nethermind.Logging;
+using Nethermind.Store;
+
+namespace Nethermind.AuRa.Validators
+{
+    public class ContractValidator : IAuRaValidatorProcessor
+    {
+        private ValidatorContract _validatorContract;
+        private readonly ILogger _logger;
+        private readonly IStateProvider _stateProvider;
+        private readonly ITransactionProcessor _transactionProcessor;
+        
+        private HashSet<Address> _validators;
+        private PendingValidators _pendingValidators;
+        private readonly ValidationStampCollection _validatedSinceFinalized = new ValidationStampCollection();
+
+        protected Address ContractAddress { get; }
+        protected IAbiEncoder AbiEncoder { get; }
+        protected long StartBlockNumber { get; }
+        protected CallOutputTracer Output { get; } = new CallOutputTracer();
+        protected ValidatorContract ValidatorContract => _validatorContract ?? (_validatorContract = CreateValidatorContract(ContractAddress));
+        private bool IsInitialized => _validators != null;
+
+        public ContractValidator(
+            AuRaParameters.Validator validator,
+            IStateProvider stateProvider,
+            IAbiEncoder abiEncoder,
+            ITransactionProcessor transactionProcessor,
+            ILogManager logManager,            
+            long startBlockNumber)
+        {
+            if (validator == null) throw new ArgumentNullException(nameof(validator));
+            if (validator.ValidatorType != Type) 
+                throw new ArgumentException("Wrong validator type.", nameof(validator));
+            
+            ContractAddress = validator.Addresses?.FirstOrDefault() ?? throw new ArgumentException("Missing contract address for AuRa validator.", nameof(validator.Addresses));
```

Comment:
why many potential addresses here and why do we always take the first then?

Reply:
is it a hacky approach to ListValidators vs SafeContract validators?

Reply:
This is in doc of that validator.Addresses property - basically for easy JSON deserialization of multiple validator types.

https://github.com/NethermindEth/nethermind/blob/647_initialize_sokol/src/Nethermind/Nethermind.Core/Specs/ChainSpecStyle/AuRaParameters.cs#L86-L93